### DISCUSSION TITLE
Optionally define a Cache folder for Assets transforms

### DISCUSSION THREAD

**Discussion by sandros94** [directus/directus#19735] [Feature Requests]
## Summary

The ability to define an optional folder that act as the `Cache` directory for the `Assets` transforms. Separating The main assets from the transformed ones.

## Basic Example

Taking as an example a local storage driver, if enabled, this feature request would by default add a `Cache` subfolder like `/directus/uploads/transforms/` and use that to store only transformed `Assets`.

## Motivation

Currently, by default, we store `Assets` and their transforms together in the same folder or bucket.

Optionally separating the two would allow to manage those transforms with external tools without ever warring of disrupting the main `Assets` that must be tracked in the `directus_files` collection. Simplifying the process of `Transforms` invalidation and/or purge.

The latter becoming particularly useful when performing full backups of a Directus instances.

### Configuration

Variable | Description | Default Value
-- | -- | --
`STORAGE_TRANSFORMS_LOCATION` | Optionally separate the transforms from the files folder. If a string with the name of a storage driver is passed that will be used. | false
`STORAGE_<LOCATION>_TRANSFORMS_ROOT` | Where to store the transforms on disk | './uploads/transforms'

### Example

Taking in example a platform that heavily depends on images as one of their main contents:

While it should be good practice to limit the number of transforms via `Presets`, something like an image provider would much prefer developers or end-user devices dictate what transform to do. On platform with moderate traffic this would easily end up with transforms surpassing the original images both in number and size. Making a purge of those a must when performing backups (either via CRONs or manually).

An additional benefit (that I still need to confirm because I have limited knowledge and usage of the S3 storage driver) could be to have a `S3` bucket set as the main `Assets` location and a `local` (or different storage driver) set as the main `Cache` target, Becoming particularly useful for S3 buckets with metered connections. Even more so if the original media gets copied to the `Cache` folder and subsequent transforms based on that copy, providing an additional predictability on S3 bills.

### Must Have:
- `STORAGE_TRANSFORMS_LOCATION` env variable to enable this feature and optionally select a storage driver as the target one.
- `STORAGE_<LOCATION>_TRANSFORMS_ROOT` env variable to define the folder location within the selected storage driver.

### Could Have:
- Expose the transforms folder to Flows, SDK and Directus App to enable deletion of the transformed assets. Like CRONs to purge or invalidate after a set `maxAge` has expired or a forced total purge.

## Drawbacks

Since this is an optional feature that is disabled by default it should not impact current Directus instances. And only a basic Guide should be written for those interested to hop in (mainly focused on the deletion of the transforms located in the assets folder).

As of right now I cannot imagine other potential drawbacks. I would be happy to discuss any in the comments.

## Alternatives

- When performing manual backups a regex becomes particularly useful to purge the transforms, but this is not only an additional step to do but a potential performance hit on larger datasets. Potentially slowing down other services. Not to mention permission concerns.

## Adoption Strategy

Being an optional feature, disabled by default, current Directus instances shouldn't be affected.

For those interested in hopping in, they could be interested in guides to purge the old transforms and/or configure different storage drivers for different aspects (main/cache).

Since the enablement of this feature requires a Directus restart this shouldn't cause additional runtime problems.

## Unresolved Questions

- Is it even possible to separate the transforms `Cache` from the main `Assets` in different storage drivers? (local/S3/etc.)
- Should it be possible to define multiple storage drivers as `Cache` targets?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
